### PR TITLE
Enable Building Rotation, Movement, and a Water Boundary Condition

### PR DESCRIPTION
**Pull request by Myoldmopar**
Pull request overview
---------------------
For a separate project, we had to alter EnergyPlus to allow for building rotation, movement, and a new water boundary condition.  I'd like to start the conversation about how this could be brought into the mainline E+ without disturbing any workflows.  Don't worry too much about the actual code already done here, there may be much better ways to do the things I've done, though through our testing we found this did work pretty well, and validated well against experimental measurements.

### CODE REVIEW

**Comment by nrel-bot-3** [comment]
@Myoldmopar @lgentile it has been 14 days since this pull request was last updated.

**Comment by nrel-bot-3** [comment]
@Myoldmopar @lgentile it has been 15 days since this pull request was last updated.

**Comment by nrel-bot-3** [comment]
@Myoldmopar @lgentile it has been 16 days since this pull request was last updated.

File: idd/Energy+.idd.in [hunk 1]
```diff
@@ -16170,15 +16196,41 @@ SurfaceProperty:OtherSideConditionsModel,
...
       \note GroundCoupledSurface provides boundary conditions for surfaces
       \note  in contact with GroundDomain objects
+      \note ConvectiveUnderwater provides a connection between a surface and
+      \note  an underwater boundary condition defined using a
+      \note  SurfaceProperty:Underwater object
       \default GapConvectionRadiation
 
+SurfaceProperty:Underwater,
+       \memo This object sets up a convective boundary condition for a surface
+       \memo The free stream velocity is determined from the SiteLocation:Varying object
```
Comment (Myoldmopar) [inline idd/Energy+.idd.in]: ToDo: Change object name

File: src/EnergyPlus/DataSurfaces.hh [hunk 2]
```diff
@@ -670,6 +670,8 @@ namespace DataSurfaces {
 		bool MovInsulIntPresent; // True when movable insulation is present
 		bool MovInsulIntPresentPrevTS; // True when movable insulation was present during the previous time step
 		// Vertices
+		//Array1D< Vector > Vertex; // Surface Vertices are represented by Number of Sides and Vector (type)
+		Array1D< Vector > NewVertex;
```
Comment (Myoldmopar) [inline src/EnergyPlus/DataSurfaces.hh]: ToDo: Remove commented new line

File: src/EnergyPlus/WeatherManager.cc [hunk 3]
```diff
@@ -586,32 +593,147 @@ namespace WeatherManager {
...
 
-		// SUBROUTINE LOCAL VARIABLE DECLARATIONS:
-		// na
+	}
 
-		Envrn = 0;
+	bool
+	CheckIfAnyUnderwaterBoundaries()
+	{
+	    bool errorsFound = false;
```
Comment (Myoldmopar) [inline src/EnergyPlus/WeatherManager.cc]: ToDo: indentation in several spots here

**Review by Myoldmopar — COMMENTED** [review]
ToDo:
 - [X] All the things I just mentioned
 - [ ] Unit tests
 - [X] Documentation
 - [X] Change IDFs to new structure

File: doc/engineering-reference/src/advanced-surface-concepts/surfaceproperty-underwater.tex [hunk 4]
```diff
@@ -0,0 +1,13 @@
+\section{Underwater Surface Boundary Condition}\label{otherside-conditions-model-underwater}
+
+\subsection{Overview}\label{overview-018}
```
Comment (Myoldmopar) [inline doc/engineering-reference/src/advanced-surface-concepts/surfaceproperty-underwater.tex:3]: This section needs to better describe the convection calculation, with a reference.

File: doc/engineering-reference/src/climate-sky-and-solar-shading-calculations/climate-calculations.tex [hunk 5]
```diff
@@ -243,6 +243,13 @@ \subsection{Weather File Solar Interpolation}\label{weather-file-solar-interpola
 
 The solar values on the weather file are average values over the hour. For interpolation of hourly weather data (i.e., when the specified timestep is greater than 1), the average value is assumed to be the value at the midpoint of the hour.~ The reported values in the output are totals for each reporting period. So, hourly reported values will not match the original values in the weather file, but the total solar for a day should agree. Reference (Ellis, Liesen, Pedersen, 2003)
 
+\subsection{Variable Location and Orientation}\label{variable-location-orientation}
+
+EnergyPlus allows for a user to specify that the building is not static for the entire simulation.
+Use of the ``Site:VariableLocation'' object allows for specifying a scheduled building location (by latitude and longitude) and orientation.
+Varying the orientation transforms the vertex locations during a simulation to allow appropriate incident solar on the rotated surfaces.
+Varying the location adjusts the solar calculation terms to account for the changed location.
```
Comment (Myoldmopar) [inline doc/engineering-reference/src/climate-sky-and-solar-shading-calculations/climate-calculations.tex:251]: This section needs to better describe the underlying mathematical model changes being made.

Comment (Myoldmopar) [reply doc/engineering-reference/src/climate-sky-and-solar-shading-calculations/climate-calculations.tex:251]: As well as how it affects other transformations (global North, other coordinate systems, appendix G rotation)

**Comment by Myoldmopar** [comment]
This was approved on the sizing call.  I'll let CI get after it a bit this evening and if it looks good I'll merge it in.  I am opening up an issue to polish up the documentation after the freeze.

**Comment by Myoldmopar** [comment]
Bah, this ended up with one build warning because it didn't have the goofy unused expression result fix in it from upstream.  If the other platforms are similar later I'll merge it in.

**Comment by Myoldmopar** [comment]
Well that looks pretty dang green.  Merging now.

**Pull request merged**